lib/Sema/CSDiag.cpp

```diff
@@ -3149,9 +3150,17 @@ void ConstraintSystem::diagnoseAssignmentFailure(Expr *dest, Type destTy,
     diagID = diag::assignment_bang_has_immutable_subcomponent;
   else if (isa<UnresolvedDotExpr>(dest) || isa<MemberRefExpr>(dest))
     diagID = diag::assignment_lhs_is_immutable_property;
-  else if (isa<SubscriptExpr>(dest))
+  else if (auto sub = dyn_cast<SubscriptExpr>(dest)) {
     diagID = diag::assignment_subscript_has_immutable_base;
-  else {
+    
+    // If the destination is a subscript with a 'dynamicLookup:' label and if
+    // the tuple is implicit, then this was actually a @dynamicMemberLookup
+    // access. Emit a more specific diagnostic.
+    if (sub->getIndex()->isImplicit() &&
+        sub->getArgumentLabels().size() == 1 &&
+        sub->getArgumentLabels().front().str() == "dynamicMember")
```

Comment:
Maybe add `dynamicMember` to `KnownIdentifiers.def`, so you don't have to hardcode the string repeatedly?

Reply:
I can do that if you'd prefer, but is it better in practice?  We'd still be hardcoding the string repeatedly, just in a different way/place.

Reply:
I'm happy to do it, and not arguing, just seek to understand why it is better.

Reply:
Mostly I think it's an anti-pattern to have the same string literal repeated in many places. Even a #define would be better, but since you're always working with identifiers, `KnownIdentifiers.def` seems the best place.